File: javademo/src/main/java/cn/shanyaliux/javademo/MainActivity.java
package cn.shanyaliux.javademo;

import androidx.appcompat.app.AppCompatActivity;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothDevice;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function2;
import world.shanya.serialport.SerialPort;
import world.shanya.serialport.SerialPortBuilder;

@SuppressLint("MissingPermission")
public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        StringBuilder stringBuilder = new StringBuilder();
        TextView textViewReceived = findViewById(R.id.textViewReceiced);
        TextView textViewConnectInfo = findViewById(R.id.textViewConnectInfo);
        Button buttonConnect = findViewById(R.id.buttonConnect);
        Button buttonDisconnect = findViewById(R.id.buttonDisconnect);
        EditText editTextSendData = findViewById(R.id.editTextTextSend);
        Button buttonSend = findViewById(R.id.buttonSend);

        SerialPort serialPort = SerialPortBuilder.INSTANCE
                .setReceivedDataCallback((Function1<String, Unit>) s -> {
                    runOnUiThread(() -> {
                        stringBuilder.append(s);
                        textViewReceived.setText(stringBuilder.toString());
                    });
                    return null;
                })
                .setConnectionStatusCallback((aBoolean, bluetoothDevice) -> {
                    runOnUiThread(() -> {
                        String info;
                        if (aBoolean) {
                            info = "设备名称:\t" + bluetoothDevice.getName() +
                                    "\n设备地址:\t" + bluetoothDevice.getAddress() +
                                    "\n 设备类型:\t" + bluetoothDevice.getType();

                        }else{
                            info = "";
                        }
                        textViewConnectInfo.setText(info);
                    });
                    return null;
                })
                .build(this);

        buttonConnect.setOnClickListener((v) -> serialPort.openDiscoveryActivity());

        buttonDisconnect.setOnClickListener((v)-> serialPort.disconnect());

        buttonSend.setOnClickListener((v-> serialPort.sendData(editTextSendData.getText().toString())));

    }
}

File: kotlindemo/src/main/java/cn/shanyaliux/kotlindemo/MainActivity.kt
package cn.shanyaliux.kotlindemo

import android.annotation.SuppressLint
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import org.w3c.dom.Text
import world.shanya.serialport.SerialPortBuilder

class MainActivity : AppCompatActivity() {
    @SuppressLint("MissingPermission", "SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val stringBuilder = StringBuilder()
        val textViewReceived = findViewById<TextView>(R.id.textViewReceiced)
        val textViewConnectInfo = findViewById<TextView>(R.id.textViewConnectInfo)
        val buttonConnect = findViewById<Button>(R.id.buttonConnect)
        val buttonDisconnect = findViewById<Button>(R.id.buttonDisconnect)
        val editTextSendData = findViewById<EditText>(R.id.editTextTextSend)
        val buttonSend = findViewById<Button>(R.id.buttonSend)

        val serialPort = SerialPortBuilder
            .setReceivedDataCallback {
                MainScope().launch {
                    stringBuilder.append(it)
                    textViewReceived.text = stringBuilder.toString()
                }
            }
            .setConnectionStatusCallback { status, bluetoothDevice ->
                MainScope().launch {
                    if (status) {
                        textViewConnectInfo.text =
                            "设备名称:\t${bluetoothDevice?.name}\n" +
                                    "设备地址:\t${bluetoothDevice?.address}\n" +
                                    "设备类型:\t${bluetoothDevice?.type}"

                    }else {
                        textViewConnectInfo.text = ""
                    }
                }
            }
            .build(this)

        buttonConnect.setOnClickListener {
            serialPort.openDiscoveryActivity()
        }

        buttonDisconnect.setOnClickListener {
            serialPort.disconnect()
        }

        buttonSend.setOnClickListener {
            serialPort.sendData(editTextSendData.text.toString())
        }
    }
}